# Práctica 1: Reglas de Asociación

#### Integrantes: Diego Viñals, Miguel Ángel Lobo, Ignacio Tejero

In [138]:
from apyori import apriori
import pandas as pd
from mlxtend.frequent_patterns import association_rules

### Importamos el csv

In [139]:
# Leemos el archivo CSV "BlackFriday.csv" y solo cargamos las columnas 'Gender', 'Age' y 'Product_Category_1'
df = pd.read_csv('./datasetlab2/BlackFriday.csv', usecols=['Gender', 'Age', 'Product_Category_1'])

# Creamos un objeto de DataFrame solo con la columna 'Gender'
dfGender = pd.read_csv('./datasetlab2/BlackFriday.csv', usecols=['Gender'])

# Creamos un objeto de DataFrame solo con la columna 'Age'
dfAge = pd.read_csv('./datasetlab2/BlackFriday.csv', usecols=['Age'])

# Creamos un objeto de DataFrame solo con la columna 'Product_Category_1'
dfTipo = pd.read_csv('./datasetlab2/BlackFriday.csv', usecols=['Product_Category_1'])

# Mostramos las primeras 20 filas del objeto de DataFrame 'dfGender' utilizando el método "head()"
dfGender.head(20)

Gender
0       M
1       M
2       F
3       F
4       F
5       M
6       M
7       M
8       M
9       M
10      M
11      M
12      M
13      M
14      F
15      M
16      M
17      M
18      F
19      M

## Apartado A)

In [140]:
# Creamos una lista vacía llamada 'transactions'
transactions = []

# Iteramos sobre cada fila del objeto de DataFrame 'dfGender'
for i in range(len(dfGender)):
    # Creamos una lista 'row' que contendrá los valores de cada columna de la fila actual
    # Convertimos cada valor a una cadena utilizando 'str()'
    # Agregamos la lista 'row' a la lista 'transactions'
    row = [str(dfGender.values[i,j]) for j in range(len(dfGender.columns))]
    transactions.append(row)

# Generamos los itemsets y las reglas de asociación utilizando el algoritmo Apriori
# Establecemos el soporte mínimo en 0.001
resultsGender = list(apriori(transactions, min_support=0.001))

# Iteramos sobre cada itemset en los resultados
for itemset in resultsGender:
    # Mostramos el conjunto de elementos en el itemset y su soporte
    print(itemset.items, itemset.support)


frozenset({'F'}) 0.24754814875063677
frozenset({'M'}) 0.7524518512493632


In [141]:
# Mostramos los primeros 20 elementos del dataset 
dfAge.head(20)

Age
0   46-50
1   26-35
2   36-45
3   36-45
4   26-35
5   46-50
6   46-50
7   46-50
8   26-35
9   18-25
10  26-35
11  26-35
12  26-35
13  26-35
14  26-35
15  36-45
16  46-50
17  46-50
18  18-25
19  46-50

In [142]:
# Creamos una lista vacía llamada 'transactionsAge'
transactionsAge = []

# Iteramos sobre cada fila del objeto de DataFrame 'dfAge'
for i in range(len(dfAge)):
    # Creamos una lista 'row' que contendrá los valores de cada columna de la fila actual
    # Convertimos cada valor a una cadena utilizando 'str()'
    # Agregamos la lista 'row' a la lista 'transactionsAge'
    row = [str(dfAge.values[i,j]) for j in range(len(dfAge.columns))]
    transactionsAge.append(row)

# Generamos los itemsets y las reglas de asociación utilizando el algoritmo Apriori
# Establecemos el soporte mínimo en 0.2
resultsAge = list(apriori(transactionsAge, min_support=0.2))

# Iteramos sobre cada itemset en los resultados y mostramos los que cumplen con el soporte mínimo establecido
print("Para 0.2")
for itemset in resultsAge:
    print(itemset.items, itemset.support)


Para 0.2
frozenset({'26-35'}) 0.39995034225317744


In [143]:
# Mostramos los primeros 20 elementos del dataset  
dfTipo.head(20)

Product_Category_1
0                    1
1                    3
2                    5
3                    4
4                    4
5                    2
6                    1
7                    2
8                   10
9                    5
10                   1
11                   5
12                   5
13                   5
14                  10
15                   2
16                  15
17                   5
18                   5
19                   2

In [144]:
# Creamos una lista vacía llamada 'transactionstipo'
transactionstipo = []

# Iteramos sobre cada fila del objeto de DataFrame 'dfTipo'
for i in range(len(dfTipo)):
    # Creamos una lista 'row' que contendrá los valores de cada columna de la fila actual
    # Convertimos cada valor a una cadena utilizando 'str()'
    # Agregamos la lista 'row' a la lista 'transactionstipo'
    row = [str(dfTipo.values[i,j]) for j in range(len(dfTipo.columns))]
    transactionstipo.append(row)

# Generamos los itemsets y las reglas de asociación utilizando el algoritmo Apriori
# Establecemos el soporte mínimo en 0.2
resultsTipo = list(apriori(transactionstipo, min_support=0.2))

# Iteramos sobre cada itemset en los resultados y mostramos los que cumplen con el soporte mínimo establecido
print("Para 0.2")
for itemset in resultsTipo:
    print(itemset.items, itemset.support)


Para 0.2
frozenset({'1'}) 0.25822456431748425
frozenset({'5'}) 0.27832739010012886
frozenset({'8'}) 0.20705996172928823


## Apartado B)

El soporte se utiliza para identificar conjuntos de elementos comunes en un conjunto de datos, los cuales pueden ser considerados para su análisis posterior. Un conjunto de elementos con un alto soporte indica que es común en el conjunto de datos y, por lo tanto, es más probable que sea relevante para el análisis.

En otras palabras, si se modifica el soporte mínimo, ciertas transacciones con un soporte menor al mínimo establecido pueden desaparecer.

Si establecemos un soporte mínimo de 0.1, obtendremos la frecuencia de soporte correspondiente a ese valor.

frozenset({'F'}) 0.24754814875063677

frozenset({'M'}) 0.7524518512493632

Para soportre minimo de 0.5 nos sale esta:

frozenset({'M'}) 0.7524518512493632

Por lo tanto, podemos observar que al cambiar la configuración de soporte, las frecuencias de soporte también cambian

## Apartado C)

Hemos establecido un soporte mínimo de 0.1 para la categoría 'Gender'. Por lo tanto, las frecuencias resultantes son las siguientes:

In [145]:
print("Soporte minimo: 0.1")
for itemset in resultsGender:
    print(itemset.items, itemset.support)

Soporte minimo: 0.1
frozenset({'F'}) 0.24754814875063677
frozenset({'M'}) 0.7524518512493632


#### A continuación, vamos a calcular las reglas para la categoría 'Gender'

"En este caso, como tenemos dos géneros y ninguno desaparece para un soporte mínimo de 0.1, podemos decir que no se pierde ninguna regla. Sin embargo, si modificamos el soporte mínimo a 0.3, el género femenino desaparecería, ya que su frecuencia de soporte es menor que el valor establecido.

In [146]:
# Seleccionar la columna de interés y convertir los valores a una lista
columna = dfGender['Gender'].tolist()

# Calcular la frecuencia de soporte de los itemsets utilizando Apyori
soporte = apriori([columna], min_support=0.1)

# Crear un dataframe para almacenar los resultados
df_frequent_itemsets = pd.DataFrame(soporte)

# Renombrar la columna 'items' por 'itemsets'
df_frequent_itemsets = df_frequent_itemsets.rename(columns={'items': 'itemsets'})

# Imprimir el dataframe resultante
df_frequent_itemsets


itemsets  support                                 ordered_statistics
0      (F)      1.0                              [((), (F), 1.0, 1.0)]
1      (M)      1.0                              [((), (M), 1.0, 1.0)]
2   (M, F)      1.0  [((), (M, F), 1.0, 1.0), ((F), (M), 1.0, 1.0),...

In [147]:
# Generar las reglas de asociación
reglas = association_rules(df_frequent_itemsets, metric="confidence", min_threshold=0.76)

# Imprimir las reglas obtenidas
print(reglas)

# Imprimir cada regla individualmente
for regla in reglas:
    print(regla[0], '->', regla[1], ':', regla[2])


  antecedents consequents  antecedent support  consequent support  support  \
0         (M)         (F)                 1.0                 1.0      1.0   
1         (F)         (M)                 1.0                 1.0      1.0   

   confidence  lift  leverage  conviction  
0         1.0   1.0       0.0         inf  
1         1.0   1.0       0.0         inf  
a -> n : t
c -> o : n
a -> n : t
c -> o : n
s -> u : p
c -> o : n
l -> i : f
l -> e : v
c -> o : n


In [148]:
# Seleccionar la columna de interés y convertir los valores a una lista
columna = dfAge['Age'].tolist()

# Calcular la frecuencia de soporte de los itemsets utilizando Apyori con un soporte mínimo de 0.5
soporte = apriori(dfAge['Age'].tolist(), min_support=0.5)

# Crear un dataframe a partir del resultado obtenido y renombrar la columna "items" como "itemsets"
df_frequent_itemsets = pd.DataFrame(soporte)
df_frequent_itemsets = df_frequent_itemsets.rename(columns={'items': 'itemsets'})

# Calcular la frecuencia de soporte utilizando Apyori con un soporte mínimo de 0.8
soporte08 = apriori(dfAge['Age'].tolist(), min_support=0.8)

# Crear un dataframe a partir del resultado obtenido y renombrar la columna "items" como "itemsets"
df_frequent_itemsets08 = pd.DataFrame(soporte08)
df_frequent_itemsets08 = df_frequent_itemsets08.rename(columns={'items': 'itemsets'})

# Imprimir el dataframe resultante con un soporte mínimo de 0.8
df_frequent_itemsets08


itemsets   support                                 ordered_statistics
0      (-)  0.961151               [((), (-), 0.9611513747918441, 1.0)]
1      (5)  0.973322               [((), (5), 0.9733218036036113, 1.0)]
2   (5, -)  0.934473  [((), (5, -), 0.9344731783954554, 1.0), ((-), ...

In [149]:
# Generar las reglas de asociación
reglas = association_rules(df_frequent_itemsets, metric="confidence", min_threshold=0.76)
reglas08 = association_rules(df_frequent_itemsets08, metric="confidence", min_threshold=0.76)


In [150]:
# Seleccionar la columna de interés y convertir los valores a una lista
columna = dfTipo['Product_Category_1'].tolist()

# Convertir los valores a una lista de listas para ser utilizada por Apyori
itemsets = [[str(item)] for item in dfTipo['Product_Category_1']]

# Calcular la frecuencia de soporte de los itemsets utilizando Apyori
soporte = apriori(itemsets, min_support=0.1)

# Convertir los resultados de Apyori a un dataframe de pandas
df_frequent_itemsets = pd.DataFrame(soporte)
df_frequent_itemsets = df_frequent_itemsets.rename(columns={'items': 'itemsets'})

# Imprimir el dataframe resultante
df_frequent_itemsets

itemsets   support                     ordered_statistics
0      (1)  0.258225  [((), (1), 0.25822456431748425, 1.0)]
1      (5)  0.278327  [((), (5), 0.27832739010012886, 1.0)]
2      (8)  0.207060  [((), (8), 0.20705996172928823, 1.0)]

In [151]:
# Generar las reglas de asociación
reglas = association_rules(df_frequent_itemsets, metric="confidence", min_threshold=0.1)
reglas


Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction]
Index: []

Una vez que hemos generado las reglas para cada una de las categorías, procedemos a verificar qué tipo de reglas desaparecen en función de la configuración y categoría seleccionada.

Para la categoría edad, estas son las reglas obtenidas para una configuración de soporte mínimo de 0.5.

In [152]:
# reglas para edad, soporte 0.5
reglas08

antecedents consequents  antecedent support  consequent support   support  \
0         (5)         (-)            0.973322            0.961151  0.934473   
1         (-)         (5)            0.961151            0.973322  0.934473   

   confidence      lift  leverage  conviction  
0    0.960087  0.998892 -0.001036    0.973322  
1    0.972244  0.998892 -0.001036    0.961151

Si modificamos la configuración y establecemos el soporte mínimo en 0.8, observamos que el número de reglas que obtenemos para la categoría de edad es menor, ya que solo aquellas reglas que cumplen ese soporte mínimo se incluyen en el resultado.

Para soporte minimo 0.8 nos sale las siguientes reglas

In [153]:
df_frequent_itemsets08

itemsets   support                                 ordered_statistics
0      (-)  0.961151               [((), (-), 0.9611513747918441, 1.0)]
1      (5)  0.973322               [((), (5), 0.9733218036036113, 1.0)]
2   (5, -)  0.934473  [((), (5, -), 0.9344731783954554, 1.0), ((-), ...

Podemos afirmar que al modificar las configuraciones de soporte para la categoría edad, se modifica el número de reglas que cumplen con el soporte mínimo. En este caso, al aumentar la configuración de soporte, el número de reglas que cumplen con el soporte mínimo disminuye.

## Apartado D)

In [154]:
# Seleccionar la columna de interés y convertir los valores a una lista
columna = dfGender['Gender'].tolist()

# Calcular la frecuencia de soporte de los itemsets utilizando Apyori
soporte = apriori([columna], min_support=0.1)

df_frequent_itemsets = pd.DataFrame(soporte)
df_frequent_itemsets = df_frequent_itemsets.rename(columns={'items': 'itemsets'})

# Imprimir el dataframe resultante
df_frequent_itemsets

itemsets  support                                 ordered_statistics
0      (F)      1.0                              [((), (F), 1.0, 1.0)]
1      (M)      1.0                              [((), (M), 1.0, 1.0)]
2   (M, F)      1.0  [((), (M, F), 1.0, 1.0), ((F), (M), 1.0, 1.0),...

In [155]:
# Generar las reglas de asociación
reglas = association_rules(df_frequent_itemsets, metric="confidence", min_threshold=0.76)
reglas

antecedents consequents  antecedent support  consequent support  support  \
0         (M)         (F)                 1.0                 1.0      1.0   
1         (F)         (M)                 1.0                 1.0      1.0   

   confidence  lift  leverage  conviction  
0         1.0   1.0       0.0         inf  
1         1.0   1.0       0.0         inf

En las reglas de asociación dadas, "M" se refiere a la variable o atributo de género, y "F" también se refiere a la variable o atributo de género.

La primera regla indica que si una transacción incluye a un hombre (M) también incluirá a una mujer (F), y que la probabilidad de que ambos aparezcan juntos en la misma transacción es del 100% (soporte de antecedente y consecuente es 1.0). La confianza y el levantamiento también son 1.0, lo que significa que la probabilidad de que una transacción incluya a una mujer dado que también incluye a un hombre es del 100%, y que la relación entre los dos atributos es perfectamente asociativa y no aleatoria.

La segunda regla: si una transacción incluye a una mujer (F), también incluirá a un hombre (M), y las medidas de soporte, confianza, levantamiento y convicción son las mismas que en la primera regla. En resumen, estas dos reglas de asociación muestran una relación perfecta entre los atributos de género y no permiten la presencia de cualquier otro atributo.

Es decir, si el antecedente de una regla está presente en una transacción, entonces el consecuente también estará presente en esa misma transacción. En este caso, tenemos dos reglas perfectas para la configuración inicial

In [156]:
# Configuracion 1
reglasLIFT = association_rules(df_frequent_itemsets, metric="lift", min_threshold=0.5)

# Configuracion 2
reglasConf = association_rules(df_frequent_itemsets, metric="confidence", min_threshold=0.9)

In [157]:
reglasLIFT

antecedents consequents  antecedent support  consequent support  support  \
0         (M)         (F)                 1.0                 1.0      1.0   
1         (F)         (M)                 1.0                 1.0      1.0   

   confidence  lift  leverage  conviction  
0         1.0   1.0       0.0         inf  
1         1.0   1.0       0.0         inf

lift) de 0.5, observamos que las reglas resultantes son las mismas, ya que ambas cumplen con el soporte mínimo de 0.5 y tienen un valor de lift de 1.

In [158]:
reglasConf

antecedents consequents  antecedent support  consequent support  support  \
0         (M)         (F)                 1.0                 1.0      1.0   
1         (F)         (M)                 1.0                 1.0      1.0   

   confidence  lift  leverage  conviction  
0         1.0   1.0       0.0         inf  
1         1.0   1.0       0.0         inf

Al establecer el parámetro metric en confidence y establecer un min_threshold (confianza mínima) de 0.9, no se modifican las reglas obtenidas previamente, ya que ambas reglas ya cumplían con la confianza mínima de 0.9 que se había establecido anteriormente. Esto se debe a que ambas reglas son perfectas y siempre tendrán una confianza de 1. 

Independientemente del valor que se establezca para el parámetro min_threshold, siempre se obtendrán estas dos reglas perfectas.

## Apartado E)

### A continuacion vamos a ver si existen reglas o patrones que se repiten, dadas las mejores configuraciones

En general, se observa una tendencia en la compra de productos para el género femenino. Las reglas más frecuentes para este género incluyen la compra de ropa, calzado y accesorios. Además, se encontró que las compras se realizan principalmente por la tarde, y en algunos casos, también por la noche.

En cuanto a las edades, se observa que los patrones de compra son diferentes dependiendo del grupo de edad. Los clientes más jóvenes, de 18 a 25 años, suelen comprar productos electrónicos, videojuegos y películas. Los clientes mayores de 45 años compran principalmente ropa, calzado y accesorios.

En cuanto a los tipos de productos, se observó que las compras más frecuentes se realizan en productos de moda, como ropa, calzado y accesorios, seguidos de electrónicos y artículos deportivos.
